sigmf/sigmffile.py

```diff
@@ -751,23 +753,58 @@ def verify_stream_hashes(self):
             if path.isfile(metafile_name):
                 new_hash = sigmf_hash.calculate_sha512(filename=metafile_name)
                 if old_hash != new_hash:
-                    raise SigMFFileError(f'Calculated file hash for {metafile_name} does not match collection metadata.')
+                    raise SigMFFileError('Calculated file hash for metadata '
+                                         f'file {metafile_name} does not '
+                                         'match collection metadata.')
+            sigmffile = [x for x in self.sigmffiles
+                         if x.name == stream.get('name')][0]
+            sigmffile_meta = sigmffile.dumps()
+            sigmffile_bytes = sigmffile_meta.encode('utf-8')
+            size_of_meta = len(sigmffile_bytes)
+            sigmffile_hash = sigmf_hash.calculate_sha512(
+                fileobj=BytesIO(sigmffile_bytes),
+                offset_and_size=(0, size_of_meta)
+            )
+            if old_hash != sigmffile_hash:
+                raise SigMFFileError('Calculated file hash for SigMFFile '
+                                     f'{sigmffile.name} does not match '
+                                     'collection metadata.')
 
     def set_streams(self, metafiles):
         '''
         configures the collection `core:streams` field from the specified list of metafiles
         '''
-        self.metafiles = metafiles
         streams = []
-        for metafile in self.metafiles:
-            if metafile.endswith('.sigmf-meta') and path.isfile(metafile):
-                stream = {
-                    "name": get_sigmf_filenames(metafile)['base_fn'],
-                    "hash": sigmf_hash.calculate_sha512(filename=metafile)
-                }
-                streams.append(stream)
-            else:
-                raise SigMFFileError(f'Specifed stream file {metafile} is not a valid SigMF Metadata file')
+        sigmffile_names = []
+        self.sigmffiles = []
+        if isinstance(metafiles[0], SigMFFile):
```

Comment:
This check could be made more rigorous, to handle the (probably unlikely) case where the metafiles list is a mixture of SigMFFiles and strings of their names

Reply:
See changes in 46e7d8f. Made it so each element is handled based on type.